pkg/archive/archive.go

```
@@ -856,179 +856,189 @@ func Tar(path string, compression Compression) (io.ReadCloser, error) {
-							// found a match - so can't skip this dir
-							return nil
-						}
-					}
+				relFilePath = strings.Replace(relFilePath, include, replacement, 1)
+			}
 
-					// No matching exclusion dir so just skip dir
-					return filepath.SkipDir
+			if err := ta.addFile(filePath, relFilePath); err != nil {
+				logrus.Errorf("Can't add file %s to tar: %s", filePath, err)
+				// if pipe is broken, stop writing tar stream to it
+				if err == io.ErrClosedPipe {
+					return err
 				}
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+	}
+	return ta.TarWriter.Close()
+}
 
-				if seen[relFilePath] {
-					return nil
-				}
-				seen[relFilePath] = true
-
-				// Rename the base resource.
-				if rebaseName != "" {
-					var replacement string
-					if rebaseName != string(filepath.Separator) {
-						// Special case the root directory to replace with an
-						// empty string instead so that we don't end up with
-						// double slashes in the paths.
-						replacement = rebaseName
-					}
+// TarWithOptions creates an archive from the directory at `srcPath`, only including files whose relative
+// paths are included in `options.IncludeFiles` (if non-nil) or not in `options.ExcludePatterns`.
+func TarWithOptions(srcPath string, options *TarOptions) (io.ReadCloser, error) {
+	if _, err := fileutils.NewPatternMatcher(options.ExcludePatterns); err != nil {
+		return nil, err
+	}
 
-					relFilePath = strings.Replace(relFilePath, include, replacement, 1)
-				}
+	pipeReader, pipeWriter := io.Pipe()
 
-				if err := ta.addFile(filePath, relFilePath); err != nil {
-					logrus.Errorf("Can't add file %s to tar: %s", filePath, err)
-					// if pipe is broken, stop writing tar stream to it
-					if err == io.ErrClosedPipe {
-						return err
-					}
-				}
-				return nil
-			}); err != nil {
-				logrus.Errorf("%s", err)
-				return
-			}
+	compressWriter, err := CompressStream(pipeWriter, options.Compression)
```

AFAICS `compressWriter` is ~entirely local to the pipe-writing goroutine; the reading goroutine doesn’t need to know it exists.

---

It is already _doing_ the compression: `compressWriter` is a layer of indirection that does the compression within its `Write` implementation, and `compressWriter.Write` is ultimately called within the producer goroutine.

(After all, `compressWriter` is writing to `pipeWriter`; the caller of `TarWithOptions` is, most of the time, blocked on reading from `pipeReader`, and it has no opportunity to write to `pipeWriter` in the meantime.)

---

I'm not sure what your original point was. Can you rephrase it?

---

> You mean: the writing routine (which becomes TarWithOptionsTo in this PR) should also do compression?

Yes. `TarWithOptionsTo( options: { Compression: … } )` can be set, and it seems simpler to implement it than to specifically refuse the option (or, worse, to silently ignore it while `TarWithOptions` implements it).

OTOH that is similar to the `PatternMatcher` situation, where moving the initialization into the writing goroutine would delay error reporting.

---

… and would the compression interfere with your plans to use reflinks?

---

>  and would the compression interfere with your plans to use reflinks?

Compression, either in the file system, or on the tar stream obviously breaks reflink copying.

Unfortunately, the `CompressStream` function does more than compression; it also introduces buffering for the `Uncompressed` case by means of the `NewWriteCloserWrapper`. AFAICT, the wrapper will hide the `ReadFrom` method of the underlying writer.

I'm not sure why the buffering is there. The `tarWriter` has its own buffer.

---

https://github.com/moby/moby/pull/7542 claims a fairly significant performance improvement, although that’s not directly attributed to exactly these lines.

I don’t see `archive.Tar` doing all that much buffering.

---

the moby PR has a bunch of anecdotal data, along with a TarUntar benchmark. The latter looks rather unrepresentative, creating 100 4 byte files in /tmp, and then repeatedly tarring them.

I made a quick test that tars a dir on a real file system. The buffering in CompressStream makes a ~10% impact when going from ext4 to ext4,

```
$ TARDEST=$HOME/tardest.tar TARDIR=$HOME/vc/linux go test -run TarLinux -bench TarLinux 
2025/03/11 13:41:38 tarring "/home/hanwen/vc/linux"
2025/03/11 13:41:45 7.221876687s
2025/03/11 13:41:55 9.783580435s
2025/03/11 13:42:03 8.324693106s
2025/03/11 13:42:13 9.535414505s
2025/03/11 13:42:21 8.34720174s
2025/03/11 13:42:31 9.622281549s
2025/03/11 13:42:39 8.726733474s
2025/03/11 13:42:48 9.115060653s
2025/03/11 13:42:57 8.690927914s
2025/03/11 13:43:06 9.37279054s
2025/03/11 13:43:06 With 8.34720174s, without 9.535414505s
PASS
ok  	github.com/containers/storage/pkg/archive	88.748s
``` 

suprprisingly, with my local hack for zero-transport copy, the buffering (which disables zero-transport) actually makes things go faster: 

```
$ TARDEST=$HOME/.cache/tardest.tar TARDIR=$HOME/.cache/linux go test -run TarLinux -bench TarLinux 
2025/03/11 13:47:06 tarring "/home/hanwen/.cache/linux"
2025/03/11 13:47:15 9.059872899s
2025/03/11 13:47:30 14.882930582s
2025/03/11 13:47:37 6.968169459s
2025/03/11 13:47:53 16.237515011s
2025/03/11 13:48:00 7.07192259s
2025/03/11 13:48:15 15.151443769s
2025/03/11 13:48:23 7.171133439s
2025/03/11 13:48:38 15.060790794s
2025/03/11 13:48:46 8.301557066s
2025/03/11 13:49:01 14.920596331s
2025/03/11 13:49:01 With 7.171133439s, without 15.060790794s
PASS
ok  	github.com/containers/storage/pkg/archive	114.837s
```

for tarring larger files, zero copy does bring benefits, 

``` 
 $ TARDEST=$HOME/.cache/tardest.tar TARDIR=$HOME/.cache/lib64 go test -run TarLinux -bench TarLinux 
2025/03/11 13:54:09 tarring "/home/hanwen/.cache/lib64"
2025/03/11 13:54:12 2.802090087s
2025/03/11 13:54:14 1.640192869s
2025/03/11 13:54:16 2.12680278s
2025/03/11 13:54:18 1.61872927s
2025/03/11 13:54:20 2.151480276s
2025/03/11 13:54:21 1.619020427s
2025/03/11 13:54:24 2.168449473s
2025/03/11 13:54:25 1.584594405s
2025/03/11 13:54:27 2.132707088s
2025/03/11 13:54:29 1.617399592s
2025/03/11 13:54:29 With 2.151480276s, without 1.61872927s
```

it seems I'll need a more sophisticated strategy here (using zero-copy only for large files), and right now, there is no good argument for dropping the buffering from CompressStream.